### ISSUE TITLE
Document GitLab CI / add an example ".gitlab-ci.yml".

### ISSUE THREAD

**Issue by hbokh** [goreleaser/goreleaser#1022]
**Is your feature request related to a problem? Please describe.**
My problem is that I can't get GitLab CI to **BOTH** push to Git**Hub** and a Docker registry. Yes, that's not a typo: push releases to GitHub.

**Describe the solution you'd like**
Using Git**Lab** CI push releases to Git**Hub** and Docker images to a Docker registry (either `docker.io` or one within `gitlab`).

**Describe alternatives you've considered**
Not an alternative, but here's the file `.gitlab-ci.yml` that I use:

```yaml
image: docker:stable
services:
- docker:dind

stages:
- build

variables:
  GORELEASER_IMAGE: goreleaser/goreleaser:v0.106.0
  CI_REGISTRY: gitlab.example.com:4567
  ## Override the settings in CI/CD settings:
  DOCKER_REGISTRY: $CI_REGISTRY
  #DOCKER_REGISTRY: https://index.docker.io/v1/
  DOCKER_USERNAME: gitlab-ci-token
  DOCKER_PASSWORD: $CI_JOB_TOKEN

before_script:
  - docker login -u gitlab-ci-token -p $CI_JOB_TOKEN $CI_REGISTRY

build:
  stage: build
  script:
    - docker pull $GORELEASER_IMAGE
    - docker run --rm --privileged --entrypoint "/bin/goreleaser" -v $PWD:/go/src/github.com/somename/goproject -v /var/run/docker.sock:/var/run/docker.sock -w /go/src/github.com/somename/goproject -e GITHUB_TOKEN -e DOCKER_USERNAME -e DOCKER_PASSWORD -e DOCKER_REGISTRY $GORELEASER_IMAGE release --rm-dist --debug
```

The Docker login is always OK:

```txt
$ echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" "$CI_REGISTRY" --password-stdin
WARNING! Your password will be stored unencrypted in /root/.docker/config.json.
Configure a credential helper to remove this warning. See
https://docs.docker.com/engine/reference/commandline/login/#credentials-store

Login Succeeded
```

The push **ALWAYS** fails in pushing to whatever registry I use. Maybe because it is `dind` (Docker-in-Docker)?

Failed push to official Docker registry:
```txt
         • running                   cmd=[docker push somename/goproject:v0.1.17]
   ⨯ release failed after 7.10s error=Docker images: failed to publish artifacts: failed to push docker image: 
The push refers to repository [docker.io/somename/goproject]
2b5af56c9523: Preparing
denied: requested access to the resource is denied
: exit status 1
ERROR: Job failed: exit code 1
```

Failed push to a GitLab registry:
```txt
         • running                   cmd=[docker push gitlab.example.com:4567/somename/goproject:v0.1.20]
   ⨯ release failed after 6.09s error=Docker images: failed to publish artifacts: failed to push docker image: 
The push refers to repository [gitlab.example.com:4567/somename/goproject]
fbd894576494: Preparing
denied: access forbidden
: exit status 1
ERROR: Job failed: exit code 1
```

So far I've set `skip_push: true` in `.goreleaser.yml` to make the CI job succeed, and push Docker images manually...

**Additional context**
I would like to see a working example in the section [Continuous Integration](https://goreleaser.com/ci/) for "GitLab CI".
Or, if this is a non-working option, a couple of lines that it is not supported.

**Comment by caarlos0** [comment]
hmm

I think we'll need to find a way to auth to more registries...

the problem is probably here: https://github.com/goreleaser/goreleaser/blob/master/scripts/entrypoint.sh#L3

**Comment by hbokh** [comment]
@caarlos0 Even using a **single** registry (like the official one for Docker) gives the issue in GitLab CI the way I have set it up. So maybe my CI configuration is plain wrong and hence asking if there is a working example.

**Comment by caarlos0** [comment]
hmm idk of anyone using gitlab-ci + goreleaser... maybe you can try on the gophers slack?

**Comment by hbokh** [comment]
Tested the release process on my Mac, so *outside* of any GitLab CI.
In my experience it seems to be not a GitLab CI issue, but a Docker issue instead.

I ran `goreleaser` both with the CLI release of GoReleaser (v0.106.0) and with the Docker image (also v0.106.0) on my Mac.

First I tagged the repo and also set some needed ENV variables:

```
export GITHUB_TOKEN=7723281fblablablarandomd34db33r
export DOCKER_PASSWORD=somePassword
export DOCKER_USERNAME=someUser
export DOCKER_REGISTRY=https://index.docker.io/v1
```

The Go binary works - command used:
```txt
$ goreleaser-0.106.0 release --rm-dist
```
Both GitHub gets release packages and Docker Hub 2 images.
In the same directory, using the Docker image will not push to any registry - command used:
```txt
$ docker run --rm --privileged --entrypoint "/bin/goreleaser" -v $PWD:/go/src/github.com/somename/goproject -v /var/run/docker.sock:/var/run/docker.sock -w /go/src/github.com/somename/goproject -e GITHUB_TOKEN -e DOCKER_USERNAME -e DOCKER_PASSWORD -e DOCKER_REGISTRY goreleaser/goreleaser:v0.106.0 release --rm-dist

...

denied: requested access to the resource is denied
: exit status 1
```

**Comment by caarlos0** [comment]
ohh, now I see it, you're overriding entrypoint...

` --entrypoint "/bin/goreleaser"`

you probably need to remove this, as the `docker login` step is on the `entrypoint.sh` file I linked before...

**Comment by hbokh** [comment]
Doh! That makes sense!

```txt
$ docker run --rm --privileged -v $PWD:/go/src/github.com/somename/goproject -v /var/run/docker.sock:/var/run/docker.sock -w /go/src/github.com/somename/goproject -e GITHUB_TOKEN -e DOCKER_USERNAME -e DOCKER_PASSWORD -e DOCKER_REGISTRY goreleaser/goreleaser:v0.106.0 release --rm-dist
Login to the docker...
WARNING! Using --password via the CLI is insecure. Use --password-stdin.
WARNING! Your password will be stored unencrypted in /root/.docker/config.json.
Configure a credential helper to remove this warning. See
https://docs.docker.com/engine/reference/commandline/login/#credentials-store

Login Succeeded

   • releasing using goreleaser 0.106.0...
   • loading config file       file=.goreleaser.yml
   • RUNNING BEFORE HOOKS
      • running go get -d -v ./...
   • GETTING AND VALIDATING GIT STATE
      • releasing v1.0.0-rc2, commit 277f968654b950cc17680e0efa55472801a9641b
   • PARSING TAG

--- cut --- 8< ---

   • PUBLISHING
      • S3
         • pipe skipped              error=s3 section is not configured
      • HTTP PUT
         • pipe skipped              error=put section is not configured
      • ARTIFACTORY
         • pipe skipped              error=artifactory section is not configured
      • DOCKER IMAGES
         • pushing docker image      image=somename/goproject:v1.0.0-rc2
   ⨯ release failed after 9.57s error=Docker images: failed to publish artifacts: failed to push docker image:
The push refers to repository [docker.io/somename/goproject]
599173c2a3b6: Preparing
denied: requested access to the resource is denied
: exit status 1
```
😢

PS I still think this is a **GREAT** tool!

**Comment by hbokh** [comment]
Some good news: I built a Docker image from commit ec13cb3 and named it `goreleaser/goreleaser:dev`. That one seems to work with the above used command. 🎉
Though I'm using an old-style `.goreleaser.yml` file meant for `v0.106.0`.

**Comment by hbokh** [comment]
Affirmative, this version also works in GitLab CI!
I have pushed the image as `voipgrid/goreleser:dev` and used this `.gitlab-ci.yml` (other ENV values are set in the CI/CD (masked) variables):

```yaml
image: docker:stable
services:
- docker:dind

stages:
- build

variables:
  GORELEASER_IMAGE: voipgrid/goreleaser:dev
  DOCKER_REGISTRY: https://index.docker.io/v1/

before_script:

build:
  stage: build
  script:
    - docker pull $GORELEASER_IMAGE
    - docker run --rm --privileged -v $PWD:/go/src/github.com/somename/goproject -v /var/run/docker.sock:/var/run/docker.sock -w /go/src/github.com/somename/goproject -e GITHUB_TOKEN -e DOCKER_USERNAME -e DOCKER_PASSWORD -e DOCKER_REGISTRY $GORELEASER_IMAGE release --rm-dist --debug
```

**Comment by caarlos0** [comment]
🤔 weird... well, guess it's time to cut a new release anyway :D

**Comment by caarlos0** [comment]
although, no changes on docker stuff on master 🤔

**Comment by hbokh** [comment]
> although, no changes on docker stuff on master 🤔

I had a look inside the container images.

`goreleaser/goreleaser:v0.106.0`:
```txt
$ docker run --rm --entrypoint "/bin/bash" -ti goreleaser/goreleaser:v0.106.0
bash-4.4# docker version
Client:
 Version:           18.09.1-ce
 API version:       1.39
 Go version:        go1.11.4
 Git commit:        4c52b901c6cb019f7552cd93055f9688c6538be4
 Built:             Fri Jan 11 15:41:33 2019
 OS/Arch:           linux/amd64
 Experimental:      false
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
bash-4.4# goreleaser --version

0.106.0, commit b8f590126537beb8f269f424bb63f8326f469152, built at 2019-04-17T12:44:51Z
bash-4.4# cat /etc/os-release
NAME="Alpine Linux"
ID=alpine
VERSION_ID=3.9.3
PRETTY_NAME="Alpine Linux v3.9"
HOME_URL="https://alpinelinux.org/"
BUG_REPORT_URL="https://bugs.alpinelinux.org/"
```

`voipgrid/goreleaser:dev`:
```txt
$ docker run --rm --entrypoint "/bin/bash" -ti voipgrid/goreleaser:dev
bash-4.4# docker version
Client:
 Version:           18.09.1-ce
 API version:       1.39
 Go version:        go1.11.4
 Git commit:        4c52b901c6cb019f7552cd93055f9688c6538be4
 Built:             Fri Jan 11 15:41:33 2019
 OS/Arch:           linux/amd64
 Experimental:      false
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
bash-4.4# goreleaser --version

dev, commit none, built at unknown
bash-4.4# cat /etc/os-release
NAME="Alpine Linux"
ID=alpine
VERSION_ID=3.9.4
PRETTY_NAME="Alpine Linux v3.9"
HOME_URL="https://alpinelinux.org/"
BUG_REPORT_URL="https://bugs.alpinelinux.org/"
```

Besides the dev build for `goreleaser` only minor difference between the Alpine versions...

**Comment by caarlos0** [comment]
weird... anyway, I'm releasing v0.107.0 right now, if you can test it I would be thankful :D

**Comment by hbokh** [comment]
> weird... anyway, I'm releasing v0.107.0 right now, if you can test it I would be thankful :D

FYI I will test it Monday at work.

**Comment by hbokh** [comment]
> weird... anyway, I'm releasing v0.107.0 right now, if you can test it I would be thankful :D

Had planned to test this from work next thing on Monday, but managed to do it from home.
This version works!! 🎉

Now, I could possibly fork this repo and add a little something in the docs [Continuous Integration](https://goreleaser.com/ci/), right?

**Comment by caarlos0** [comment]
awesome, yes, thanks for the PR :D

**Comment by caarlos0** [comment]
fixed on #102
